common/src/main/java/com/microsoft/identity/common/adal/internal/tokensharing/TokenShareUtility.java

```diff
@@ -0,0 +1,225 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.adal.internal.tokensharing;
+
+import android.support.annotation.NonNull;
+import android.util.Pair;
+
+import com.microsoft.identity.common.exception.BaseException;
+import com.microsoft.identity.common.exception.ClientException;
+import com.microsoft.identity.common.exception.ServiceException;
+import com.microsoft.identity.common.internal.cache.ICacheRecord;
+import com.microsoft.identity.common.internal.cache.MsalOAuth2TokenCache;
+import com.microsoft.identity.common.internal.dto.AccountRecord;
+import com.microsoft.identity.common.internal.dto.IdTokenRecord;
+import com.microsoft.identity.common.internal.dto.RefreshTokenRecord;
+import com.microsoft.identity.common.internal.logging.Logger;
+import com.microsoft.identity.common.internal.providers.microsoft.MicrosoftAccount;
+import com.microsoft.identity.common.internal.providers.microsoft.MicrosoftIdToken;
+import com.microsoft.identity.common.internal.providers.microsoft.MicrosoftRefreshToken;
+import com.microsoft.identity.common.internal.providers.microsoft.azureactivedirectory.AzureActiveDirectoryIdToken;
+import com.microsoft.identity.common.internal.providers.oauth2.IDToken;
+import com.nimbusds.jose.JOSEObjectType;
+import com.nimbusds.jose.PlainHeader;
+import com.nimbusds.jwt.JWT;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.PlainJWT;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
+
+import static com.microsoft.identity.common.exception.ClientException.TOKEN_CACHE_ITEM_NOT_FOUND;
+import static com.microsoft.identity.common.internal.migration.TokenCacheItemMigrationAdapter.renewToken;
+import static com.microsoft.identity.common.internal.migration.TokenCacheItemMigrationAdapter.sBackgroundExecutor;
+
+public class TokenShareUtility implements ITokenShareInternal {
+
+    private static final String TAG = TokenShareUtility.class.getSimpleName();
+    private static final Map<String, String> mClaimRemapper = new HashMap<>();
+
+    /*
+     * Populate v1/v2 IdToken claims mapping here.
```

Comment:
@shoatman Here's where I'm doing v2 -> v1 claim mapping. To add more mappings, we add more `put()` calls here...

So far, I have just one special case:
- `preferred_username` becomes `upn`

I suspect there's a handful of these... something related to guest/idp/iss? Anything else?

Reply:
Not clear on how `sub`, `email`, and `upn` align exactly...

Reply:
sub is a pairwise identifier... an unique identifier for the app that only a particular client id will receive.  I doubt that EVO sends a pairwise identifier for the Microsoft family... and generally the FOCI apps are not using sub that I"m aware of.  Email and UPN can be different, but often are the same.

Reply:
ADAL only falls back to subject if OID is not present.  I believe that OID is present for all accounts outside of MSA passthrough.  So likely don't need to populate sub.

Reply:
OK - sounds like this stays as-is... that is: we remap `preferred_username` to `upn` and that's it.

ACK?